src/storage/btree_load.c

```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
+	  old_page = page;
+
+	  /* Set correct VPID. */
+	  *vpid = next_vpid;
+
+	  /* Init variables. */
+	  *slot_id = -1;
+	  local_header = NULL;
+	  *key_cnt = -1;
+	}
+    }
+
+  /* Scan has not started yet or a new page is required. */
+  if (page == NULL)
+    {
+      /* Get page and fix it. */
+      page = pgbuf_fix (thread_p, vpid, OLD_PAGE, PGBUF_LATCH_READ, PGBUF_UNCONDITIONAL_LATCH);
+      if (page == NULL)
+	{
+	  ASSERT_ERROR_AND_SET (ret);
+	  return ret;
+	}
+      first_scan = true;
+
+      /* Unfix the old page, if any. */
+      if (old_page != NULL)
+	{
+	  pgbuf_unfix_and_init (thread_p, old_page);
+	}
+    }
+
+  /* Check page. */
+  (void) pgbuf_check_page_ptype (thread_p, page, PAGE_BTREE);
+
+  if (local_header == NULL)
+    {
+      /* Get the header of the page. */
+      local_header = btree_get_node_header (thread_p, page);
+    }
+
+  if (*key_cnt == -1)
+    {
+      /* Get number of keys in page. */
+      *key_cnt = btree_node_number_of_keys (thread_p, page);
+    }
+
+  /* Get the number of visible oids if its the first scan of the current leaf. */
+  if (mvcc != NULL && num_visible == -1)
+    {
+      num_visible = btree_get_number_of_items (thread_p, btid, page, mvcc);
+      if (num_visible == 0)
+	{
+	  goto start_alg;
+	}
+      if (num_visible < 0)
+	{
+	  /* Error. */
+	  ret = num_visible;
+	  goto end;
+	}
+    }
+
+  adv = is_desc ? -1 : 1;
+
+  /* Check if this is the first time we try to get a value from this page. */
+  if (*slot_id == -1)
+    {
+      *slot_id = is_desc ? *key_cnt : 1;
+    }
+
+  ret = btree_get_value_from_leaf_slot (thread_p, btid, page, *slot_id, key);
+
+  *slot_id += adv;
+
+end:
+
+  *header = local_header;
+  *pg_ptr = page;
+  return ret;
+}
+
+/*
+ *  btree_get_number_of_items():-	Gets number of visible oids from a page.
+ *
+ *  thread_p(in):		    -	Thread entry.
+ *  btid(in):			    -	B-tree info.
+ *  pg_ptr(in):			    -	Page pointer.
+ *  mvcc_snapshot(in):		    -	The MVCC snapshot.
+ *
+ *  return: Number of visible items, or error code if any error occurs.	    
+ */
+
+static int
+btree_get_number_of_items (THREAD_ENTRY * thread_p, BTID_INT * btid, PAGE_PTR pg_ptr, MVCC_SNAPSHOT * mvcc_snapshot)
+{
+  RECDES record;
+  LEAF_REC leaf;
+  int num_visible = 0;
+  int key_offset = 0;
+  int ret = NO_ERROR;
+  DB_VALUE key;
+  bool clear_key;
+
+  DB_MAKE_NULL (&key);
+
+  /* Get the record. */
+  if (spage_get_record (thread_p, pg_ptr, 1, &record, PEEK) != S_SUCCESS)
+    {
+      assert_release (false);
+      ret = ER_FAILED;
+      return ret;
+    }
+
+  /* Read the record. */
+  ret =
+    btree_read_record (thread_p, btid, pg_ptr, &record, &key,
+		       &leaf, BTREE_LEAF_NODE, &clear_key, &key_offset, PEEK_KEY_VALUE, NULL);
+
+  if (ret != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      return ret;
+    }
+
+
+  num_visible =
+    btree_get_num_visible_from_leaf_and_ovf (thread_p, btid, &record, key_offset, &leaf, NULL, mvcc_snapshot);
```

Comment:
please update btree_get_num_visible_from_leaf_and_ovf and btree_record_get_num_visible_oids to return error and output number of visible as argument.